README.txt
Introduction
============

Quintagroup Captcha Core (quintagroup.captcha.core) is a core package of simple
captcha implementation. It allows to configure captchas on your Plone site. 

This product works together with other Quintagroup captcha products.
To protect standard Plone forms with captcha - use quintagroup.plonecaptchas package.
https://pypi.python.org/pypi/quintagroup.plonecaptchas 

quintagroup.captcha.core Usage
------------------------------

After product installation you'll see 'Plone captchas setup' option under 
'Add-on Product Configuration' in Site Setup. There you can select what kind 
of captchas you want to appear: either static or dynamic. In case you select 
dynamic - captcha images will be generated on the fly and you will be able to
configure captchas look by using different font sizes, background and font 
colours, period, amplitude, random values.

Compatibility
-------------

* Plone 3.x
* Plone 4.x

Links
-----

* Documentation - http://projects.quintagroup.com/products/wiki/quintagroup.captcha.core
* SVN Repository - http://svn.quintagroup.com/products/quintagroup.captcha.core
* PyPi package - https://pypi.python.org/pypi/quintagroup.captcha.core

Notes
-----

* If you want captcha for Plone default forms - use quintagroup.plonecaptchas product http://projects.quintagroup.com/products/wiki/quintagroup.plonecaptchas

* If you want captcha support for PloneFormGen forms - use quintagroup.pfg.captcha product http://projects.quintagroup.com/products/wiki/quintagroup.pfg.captcha

Authors
-------

The product is developed and maintained by http://quintagroup.com team.

* Andriy Mylenkyy
* Volodymyr Cherepanyak

Copyright (c) "Quintagroup": http://quintagroup.com, 2004-2010



docs/HISTORY.txt
Changelog
=========

0.5 - (2014-01-21)
------------------
* Added honeypot captcha functionality [kroman0]

0.4.3 - (2013-07-04)
--------------------
* Added the backward compatibility with Python 2.4 [potar]

0.4.2 - (2013-06-26)
--------------------

* Added transifex config [kroman0]

* Updated French translations from transifex, many thanks to Marc Sokolovitch
  [kroman0]

* Added inline validation detector [vmaksymiv]

0.4.1 - (2013-02-11)
--------------------

* Fixed ValueError on validation for dynamic captcha [kroman0]

* Fixed AttributeError on validation [kroman0]

0.4 - (2013-01-17)
------------------

* Updated pt_BR translation [cleber_jsantos]

* Updated classifiers [vmaksymiv]

* Cleanup code [vmaksymiv]

* Fixed ValueError on validation [kroman0] 

0.3 - (2011-07-22)
------------------

* Plone 4.1 comopatibility release [chervol]

0.2.4 - (2010-09-08)
--------------------

* removed tabindex from configlet [chervol]

0.2.3 -(2010-09-08)
-------------------

* fixed import errors [chervol]

0.2.2 - (2010-07-19)
--------------------

* Added German translation
  [Fabian Reinhard]


0.2.1 (2010-06-17)
------------------

* Fixed captcha_widget issue
  (http://plone.org/products/plone-comments/issues/5)
  [mylan]


0.2 - June 9, 2010
------------------

* Added Plone-4 support [mylan, kroman0]
* Added italian translation [kroman0]
* Updated translations [olha, mylan]
* Added, updated tests [mylan]


0.1 - Apr 7, 2010
-----------------

* Initial release


quintagroup/captcha/core/setuphandlers.py
from random import randint

from Products.CMFCore.utils import getToolByName

from quintagroup.captcha.core.config import CAPTCHA_KEY, CONFIGLET_ID, \
    PROPERTY_SHEET


def generateKey(length):
    key = ''
    for i in range(length):
        key += str(randint(0, 9))
    return key


def setupVarious(context):
    if context.readDataFile('quintagroup.captcha.core_various.txt') is None:
        return

    site = context.getSite()

    # set captcha key
    value = generateKey(8)
    if site.hasProperty(CAPTCHA_KEY):
        site._updateProperty(CAPTCHA_KEY, value)
    else:
        site._setProperty(CAPTCHA_KEY, value, 'string')


def uninstall(context):
    # Only run step if a flag file is present (e.g. not an extension profile)
    if context.readDataFile('quintagroup.captcha.core_uninstall.txt') is None:
        return

    site = context.getSite()

    # remove configlet
    cpt = getToolByName(site, 'portal_controlpanel')
    if CONFIGLET_ID in [o.id for o in cpt.listActions()]:
        cpt.unregisterConfiglet(CONFIGLET_ID)

    # remove property sheet
    pp = getToolByName(site, 'portal_properties')
    if PROPERTY_SHEET in pp.objectIds():
        pp.manage_delObjects(ids=[PROPERTY_SHEET])

    # remove captcha key property
    if site.hasProperty(CAPTCHA_KEY):
        site._delProperty(CAPTCHA_KEY)


def add_honeypot_property(context):
    pp = getToolByName(context, 'portal_properties')
    if PROPERTY_SHEET in pp.objectIds() and not pp[PROPERTY_SHEET].hasProperty("honeypot"):
        pp[PROPERTY_SHEET].manage_addProperty("honeypot", False, 'boolean')


quintagroup/captcha/core/skins/captcha_core/dynamic/getCaptchaImage.py
## Script (Python) "getCaptchaImage"
##bind container=container
##bind context=context
##bind namespace=
##bind script=script
##bind subpath=traverse_subpath
##parameters=
##title=
from quintagroup.captcha.core.utils import gen_captcha, decrypt, getWord, parseKey
from Products.CMFCore.utils import getToolByName
import random
propTool = getToolByName(context, 'portal_properties')
captchaProps = propTool['qPloneCaptchas']

hk = context.REQUEST.traverse_subpath[0]
dk = decrypt(context.captcha_key, hk)
key = parseKey(dk)['key']

text = getWord(int(key))
size = captchaProps.getProperty('image_size')
bkground = captchaProps.getProperty('background')
font_color = captchaProps.getProperty('font_color')
kwargs = {'text': text,
          'size': size,
          'bkground': bkground,
          'font_color': font_color}
if captchaProps.getProperty('random_params', 'False'):
    period = random.uniform(0.05, 0.12)
    amplitude = random.uniform(3.0, 6.5)
else:
    period = captchaProps.getProperty('period')
    amplitude = captchaProps.getProperty('amplitude')

kwargs['distortion'] = [period, amplitude, (0.0, 0.0)]

im = gen_captcha(**kwargs)
context.REQUEST.RESPONSE.setHeader('Content-Type', 'image/jpeg')
context.REQUEST.RESPONSE.setHeader('Content-Length', im['size'])
context.REQUEST.RESPONSE.setHeader('Accept-Ranges', 'bytes')
return im['src']


quintagroup/captcha/core/utils.py
import os
import re
import math
import inspect
import operator
try:
    import hashlib as md5
    md5.md5
except ImportError:
    import md5
from string import atoi
from random import randint
from random import shuffle

from DateTime import DateTime

from plone.memoize import forever

from quintagroup.captcha.core.data import basic_english
# import quintagroup.captcha.core configuration values
from quintagroup.captcha.core.config import (
    DEFAULT_IMAGE_SIZE, DEFAULT_BG, DEFAULT_FONT_COLOR, DEFAULT_DISTORTION,
    CAPTCHAS_COUNT)

try:
    import Crypto.Cipher.DES as Crypto
    Crypto
except ImportError:
    import Crypto


@forever.memoize
def get_source_file_path(obj):
    return inspect.getsourcefile(obj)


try:
    # introduced in Python 2.5
    from __builtin__ import any
except ImportError:
    # BBB: Python 2.4 support
    def any(iterable):
        for element in iterable:
            if element:
                return True
        return False


def detectInlineValidation(validation_module):
    validation_module_path = get_source_file_path(validation_module)
    path = operator.itemgetter(1)
    return any(path(frame) == validation_module_path
               for frame in inspect.stack())


def encrypt1(s):
    return md5.md5(s).hexdigest().upper()


def getTransform(x, y, a, p, o):
    return (math.sin((y + o[0]) * p) * a + x, math.sin((x + o[1]) * p) * a + y)


def gen_captcha(**kwargs):
    """Generate a captcha image"""
    try:
        from App import ImageFile
        ImageFile
    except ImportError:
        import ImageFile
    from PIL import Image
    from PIL import ImageFont
    from PIL import ImageDraw
    import random
    from PIL import ImageFile as pyImageFile
    import sys
    sys.modules['ImageFile'] = pyImageFile
    from cStringIO import StringIO

    text = kwargs.get('text', None)
    fnt_sz = kwargs.get('size', DEFAULT_IMAGE_SIZE)
    bkground = kwargs.get('bkground', DEFAULT_BG)
    font_color = kwargs.get('font_color', DEFAULT_FONT_COLOR)
    distortion = kwargs.get('distortion', DEFAULT_DISTORTION)

    period = distortion[0]
    amplitude = distortion[1]
    offset = distortion[2]

    outFile = StringIO()

    DATA_PATH = os.path.abspath(os.path.dirname(__file__)) + '/data'
    FONT_PATH = DATA_PATH + '/fonts'

    # select font for captcha text
    ALL_FONTS = ('Bd', 'It', 'MoBI', 'Mono', 'Se',
                 'BI', 'MoBd', 'MoIt', 'SeBd', '')
    rand_font = random.choice(ALL_FONTS)
    rand_font_path = FONT_PATH + '/vera/Vera%s.ttf' % rand_font
    font = ImageFont.truetype(rand_font_path, fnt_sz)
    textSize = font.getsize(text)

#------------------------------render   background1 -----------------------
    image = Image.new('RGB', (textSize[0] + 7, textSize[1] + 7), bkground)
    image.paste(bkground)
#------------------------------render       Text2 ------------------------
    draw = ImageDraw.Draw(image)
    alignment = (random.uniform(0, 1), random.uniform(0, 1))
    x = int((image.size[0] - textSize[0]) * alignment[0] + 0.5)
    y = int((image.size[1] - textSize[1]) * alignment[1] + 0.5)

    draw.text((x, y), text, font=font, fill=font_color)

#------------------------------render       Distortion -----------------------
    r = 1
    xPoints = image.size[0] / r + 2
    yPoints = image.size[1] / r + 2

    # Create a list of arrays with transformed points
    xRows = []
    yRows = []
    for j in xrange(yPoints):
        xRow = []
        yRow = []
        for i in xrange(xPoints):
            x, y = getTransform(i * r, j * r, amplitude, period, offset)

            # Clamp the edges so we don't get black undefined areas
            x = max(0, min(image.size[0] - 1, x))
            y = max(0, min(image.size[1] - 1, y))

            xRow.append(x)
            yRow.append(y)
        xRows.append(xRow)
        yRows.append(yRow)

    # Create the mesh list, with a transformation for
    # each square between points on the grid
    mesh = []
    for j in xrange(yPoints - 1):
        for i in xrange(xPoints - 1):
            mesh.append((
                # Destination rectangle
                (i * r, j * r,
                 (i + 1) * r, (j + 1) * r),
                # Source quadrilateral
                (xRows[j][i], yRows[j][i],
                 xRows[j + 1][i], yRows[j + 1][i],
                 xRows[j + 1][i + 1], yRows[j + 1][i + 1],
                 xRows[j][i + 1], yRows[j][i + 1]),
            ))

    img = image.transform(image.size, Image.MESH, mesh, Image.BILINEAR)

    # save the image to a file
    img.save(outFile, format='jpeg')
    outFile.seek(0)
    src = outFile.read()
    size = len(src)
    sys.modules['ImageFile'] = ImageFile
    return {'src': src, 'size': size}


def getWord(index):
    words = basic_english.words.split()
    return words[index]


def getIndex(word):
    words = basic_english.words.split()
    try:
        res = words.index(word)
    except ValueError:
        res = len(words) + 1
    return res


def getCaptchasCount(dynamic):
    if dynamic:
        return len(basic_english.words.split())
    else:
        return CAPTCHAS_COUNT


def getRandIndex(dynamic):
    if dynamic:
        index = randint(0, getCaptchasCount(True) - 1)
    else:
        index = randint(1, getCaptchasCount(False))
    return index


def formKey(num):
    def normalize(s):
        return (not len(s) % 8 and s) or normalize(s + str(randint(0, 9)))

    return normalize('%s_%i_' % (str(DateTime().timeTime()), num))


def encrypt(key, s):
    return toHex(Crypto.new(key).encrypt(s))


def decrypt(key, s):
    return Crypto.new(key).decrypt(toStr(s))


def parseKey(s):
    ps = re.match('^(.+?)_(.+?)_', s)
    if ps is None:
        return {'date': '', 'key': ''}
    return {'date': ps.group(1), 'key': ps.group(2)}


def toHex(s):
    lst = []
    for ch in s:
        hv = hex(ord(ch)).replace('0x', '')
        if len(hv) == 1:
            hv = '0' + hv
        lst.append(hv)

    return reduce(lambda x, y: x + y, lst)


def toStr(s):
    return s and chr(atoi(s[:2], base=16)) + toStr(s[2:]) or ''


def encrypt_token(captcha_key, hashkey, index):
    l = len(hashkey) / 16
    i = index % l * 16
    return encrypt(captcha_key, hashkey[i:i+16])


def validate_token(captcha_key, hashkey, index, tokenkey):
    l = len(hashkey) / 16
    x = int(index) % l * 16
    key = decrypt(captcha_key, tokenkey)
    return key != hashkey[x:x+16]


def validate_key(context, test_key, index, dynamic):
    if dynamic:
        enc = test_key
        solution = getWord(int(index))
    else:
        enc = encrypt1(test_key)
        solution = getattr(context, '%s.jpg' % index).title
    return enc != solution


def obfuscate_code(hashkey, token):
    strings = [
        'getElementById', 'hashkey', 'parentNode', 'style', 'display', 'none',
        'setAttribute', 'name', 'tokenkey', 'value', 'token'
    ]
    shuffle(strings)
    keys = dict([(value, index) for index, value in enumerate(strings)])
    strings[strings.index('token')] = token
    strings[strings.index('hashkey')] = hashkey
    strings = [
        "".join([
            str(hex(ord(i))).replace('0x', '\\x')
            for i in s
        ])
        for s in strings
    ]
    keys['strings'] = '"' + '","'.join(strings) + '"'
    blocks = [
        "_i[_s[%(parentNode)s]][_s[%(style)s]][_s[%(display)s]]=_s[%(none)s]",
        "_i[_s[%(setAttribute)s]](_s[%(name)s],_s[%(tokenkey)s])",
        "_i[_s[%(value)s]]=_s[%(token)s];",
    ]
    shuffle(blocks)
    value = "var _s=[%(strings)s], _i=document[_s[%(getElementById)s]](_s[%(hashkey)s])"
    value = ";".join([value] + blocks)
    return value % keys
